File: 61-rotate-list/61-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        sanitel = ListNode()
        sanitel.next = head
        p = head
        q = head
        n = 0
        count = head
        while(count):
            count = count.next
            n = n + 1
        if n != 0:
            for i in range(k%n):
                p = p.next
            while(p.next):
                q = q.next
                p = p.next
            begin = sanitel
            for i in range(k%n):
                r = q.next
                q.next = r.next
                r.next = begin.next
                begin.next = r
                begin = begin.next
        return sanitel.next
            
        
        
            
            

File: 61-rotate-list/61-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        //calculate the length of list
        int listLen = 0;
        ListNode node = head;
        ListNode tail = head;
        if(head == null)return null;
        while(node != null){
            if(node.next == null){
                tail = node;
            }
            node = node.next;
            listLen++;
        }
        k = k%listLen;
        if(k == 0)return head;
        //find the listLen - k th node.
        ListNode dumb = new ListNode(0);
        dumb.next = head;
        ListNode p = dumb;
        int count = 0;
        while(count < listLen - k){
            p = p.next;
            count++;
        }
        ListNode second = p.next;
        p.next = null;
        
        dumb.next = second;
        tail.next = head;
        return dumb.next;
        
        
        
    }
}